extensions-core/avro-extensions/src/main/java/io/druid/data/input/avro/InlineSchemasAvroBytesDecoder.java

```
@@ -0,0 +1,113 @@
+import com.metamx.common.parsers.ParseException;
+import io.druid.guice.annotations.Json;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.util.ByteBufferInputStream;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ */
+public class InlineSchemasAvroBytesDecoder implements AvroBytesDecoder
+{
+  private static final Logger logger = new Logger(InlineSchemasAvroBytesDecoder.class);
+
+  private final Map<Integer, Schema> schemaObjs;
+  private final Map<String, Map<String, Object>> schemas;
+
+  @JsonCreator
+  public InlineSchemasAvroBytesDecoder(
+      @JacksonInject @Json ObjectMapper mapper,
+      @JsonProperty("schemas") Map<String, Map<String, Object>> schemas
+  ) throws Exception
+  {
+    Preconditions.checkArgument(
+        schemas != null && schemas.size() > 0,
+        "at least one schema must be provided in schemas attribute"
+    );
+
+    this.schemas = schemas;
+
+    schemaObjs = new HashMap<>(schemas.size());
+    for (Map.Entry<String, Map<String, Object>> e : schemas.entrySet()) {
+
+      int id = Integer.parseInt(e.getKey());
+
+      Map<String, Object> schema = e.getValue();
+      String schemaStr = mapper.writeValueAsString(schema);;
+
+      logger.info("Schema string [%s] = [%s]", id, schemaStr);
+      schemaObjs.put(id, new Schema.Parser().parse(schemaStr));
+    }
+  }
+
+  @VisibleForTesting
+  public InlineSchemasAvroBytesDecoder(
+      Map<Integer, Schema> schemaObjs
+  )
+  {
+    this.schemaObjs = schemaObjs;
+    this.schemas = null;
+  }
+
+  @JsonProperty
+  public Map<String, Map<String, Object>> getSchemas()
+  {
+    return schemas;
+  }
+
+  @Override
+  public GenericRecord parse(ByteBuffer bytes)
+  {
+    int schemaId = bytes.getInt();
```

If a record is not properly formed, for example if the `ByteBuffer` is empty, is it better to throw a parse exception?